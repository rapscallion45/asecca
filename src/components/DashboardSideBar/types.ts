/*
 * Dashboard Sidebar type definitions
 *
 * These type definitions are used in the applications Dashboard Sidebar component.
 */

/**
 * Nav Bar sidebar on close callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback IDashboardSidebarOnCloseCallback
 * @returns {void} - no return value
 */
export interface IDashboardSidebarOnCloseCallback {
  (): void;
}
