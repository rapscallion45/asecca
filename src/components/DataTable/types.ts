/*
 * Data Table type definitions
 *
 * These type definitions are used in the applications data table component.
 * Only make changes to this file if the corresponding change is made in the
 * Data Table component(s)
 */

/**
 * Data Table Row Cell Value types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof DataTable
 *
 * @typedef DataTableRowCellValue
 */
export type DataTableRowCellValue =
  | string
  | Array<string>
  | boolean
  | null
  | undefined;

/**
 * Data Table Column types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof DataTable
 *
 * @typedef DataTableColumnType
 */
export type DataTableColumnType = 'string' | 'currency' | 'select' | 'check';

/**
 * Data Table Column properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableColumn
 * @prop {string} label - data table column name
 * @prop {string} key - data table column data identifier
 * @prop {DataTableColumnType} type - data table column type
 */
export interface IDataTableColumn {
  label: string;
  key: string;
  type: DataTableColumnType;
}

/**
 * Data Table Row properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableRow
 * @prop {string} label - data table row name
 */
export interface IDataTableRow {
  label: string;
}

/**
 * Data Table Cell Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCellValueCallback
 * @param {DataTableRowCellValue} value - updated cell value, can be null
 * @param {string} colKey - cell column data identifier
 * @param {number} rowIdx - cell row index in data table
 * @returns {void} - no return value
 */
export interface IDataTableEditCellValueCallback {
  (value: DataTableRowCellValue, colKey: string, rowIdx: number): void;
}

/**
 * Data Table Cell Value Retrieval callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableGetCellValueCallback
 * @param {number} rowIdx - cell row index in data table
 * @param {IDataTableColumn} column - cell column in data table
 * @returns {DataTableRowCellValue} - cell value, can be null or undefined
 */
export interface IDataTableGetCellValueCallback {
  (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue;
}

/**
 * Data Table Currency Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCurrencyCellValueCallback
 * @param {string | null} value - updated cell value
 * @returns {void} - no return value
 */
export interface IDataTableEditCurrencyCellValueCallback {
  (value: string | null): void;
}

/**
 * Data Table Checkbox Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @callback IDataTableEditCheckboxCellValueCallback
 * @param {boolean | null} value - updated cell boolean value
 * @returns {void} - no return value
 */
export interface IDataTableEditCheckboxCellValueCallback {
  (value: boolean | null): void;
}
