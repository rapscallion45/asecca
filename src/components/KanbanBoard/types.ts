/*
 * Kanban Board type definitions
 *
 * These type definitions are used in the applications kanban board component.
 * Only make changes to this file if the corresponding change is made in the
 * Kanban Board component(s)
 */

/**
 * Kanban Board Column On Drop callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @callback IKanbanBoardColumnOnDropCallback
 * @param {number} colIndex - column index of task
 * @param {number} prevColIndex - previous column index of task
 * @param {number} taskIndex - index of task
 * @returns {void} - no return value
 */
export interface IKanbanBoardColumnOnDropCallback {
  (colIndex: number, prevColIndex: number, taskIndex: number): void;
}
