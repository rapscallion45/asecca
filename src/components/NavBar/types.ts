/*
 * Nav Bar type definitions
 *
 * These type definitions are used in the applications Nav Bar component.
 */

/**
 * Nav Bar sidebar open callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback INavBarOnSidebarOpenCallback
 * @returns {void} - no return value
 */
export interface INavBarOnSidebarOpenCallback {
  (): void;
}
