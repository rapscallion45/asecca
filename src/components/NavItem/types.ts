/*
 * Nav Item type definitions
 *
 * These type definitions are used in the applications Nav Item component.
 */

/**
 * Nav Item configuration properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof NavItem
 *
 * @typedef INavItemConfig
 * @prop {string} title - nav item title text
 * @prop {string} path - nav item URL path
 * @prop {any} icon - nav item icon
 * @prop {boolean} newTab - nav item link opens in new tab
 * @prop {Array<INavItemConfig>} children - nav item child components
 */
export interface INavItemConfig {
  title: string;
  path?: string;
  icon?: any;
  newTab?: boolean;
  children?: Array<INavItemConfig>;
}
