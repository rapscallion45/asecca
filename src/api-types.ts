/*
 ** Costs Config return data type definition
 */
export type CostsConfigData = {
  application:
    | 'Per Device'
    | 'Per Batch'
    | 'Per Collection'
    | 'Per Project'
    | 'Per Destruction';
  collection_charge?: number | null;
  cost_source: 'Global' | 'Customer' | 'Project' | 'Collection';
  customer_charge?: number | null;
  effective_charge: number | null;
  global_charge?: number | null;
  line_type: 'Typical' | 'Custom';
  name: string;
  project_charge?: number | null;
};

/*
 ** Costs Config return data payload type definition
 */
export type CostsConfigDataPayload = {
  costs: Array<CostsConfigData>;
};

/*
 ** Proxy Error return data payload type definition
 */
export type ProxyErrorPayload = {
  message: string;
};
