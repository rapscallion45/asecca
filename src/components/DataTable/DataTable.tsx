import { FC, Fragment } from 'react';
import { styled } from '@mui/material/styles';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, { tableCellClasses } from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import Box from '@mui/material/Box';
import CircularProgress from '@mui/material/CircularProgress';
import Typography from '@mui/material/Typography';
import ErrorIcon from '@mui/icons-material/Error';
import DataTableRow from './DataTableRow/DataTableRow';
import ClientOnly from '../ClientOnly/ClientOnly';

/* table head cell stylings */
const StyledTableHeadCell = styled(TableCell)(({ theme }) => ({
  /* table head colors */
  [`&.${tableCellClasses.head}`]: {
    backgroundColor: theme.palette.secondary.main,
    color: theme.palette.primary.contrastText,
    fontSize: 16,
    fontWeight: 400,
  },
}));

interface DataTableProps {
  name: string;
  editColName: string;
  columns: any;
  rows: any;
  isLoading?: boolean;
  isError?: boolean;
}

/* Data Table */
/* ========== */
const DataTable: FC<DataTableProps> = (props) => {
  const {
    name,
    editColName,
    columns,
    rows,
    isLoading = false,
    isError = false,
  } = props;

  /* ClientOnly used to not allow tables to SSR */
  return (
    <ClientOnly>
      <TableContainer component={Paper}>
        <Table aria-label={`${name} table`}>
          <TableHead>
            <TableRow>
              {/* map passed columns to table headers row */}
              {columns.map((column: any) => (
                <StyledTableHeadCell key={column.key} align="center">
                  {column.label}
                </StyledTableHeadCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            <>
              {!isLoading && !isError && (
                <>
                  {/* map passed rows */}
                  {rows.map((row: any) => (
                    <Fragment key={row.name}>
                      <DataTableRow
                        row={row}
                        columns={columns}
                        /* we need the edit col to get 'Prevailing' value */
                        editCol={columns.find(
                          (col: any) => editColName === col.label
                        )}
                      />
                    </Fragment>
                  ))}
                </>
              )}
              {!isLoading && isError && (
                <Box
                  display="flex"
                  flexDirection="column"
                  py={7}
                  alignItems="center"
                >
                  <ErrorIcon color="error" fontSize="large" />
                  <Typography variant="h6" mt={2}>
                    Error loading the requested data
                  </Typography>
                </Box>
              )}
              {isLoading && (
                <Box
                  display="flex"
                  flexDirection="column"
                  py={7}
                  alignItems="center"
                >
                  <CircularProgress />
                  <Typography variant="h6" mt={2}>
                    Loading...
                  </Typography>
                </Box>
              )}
            </>
          </TableBody>
        </Table>
      </TableContainer>
    </ClientOnly>
  );
};

export default DataTable;
