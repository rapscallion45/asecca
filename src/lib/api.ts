import base64 from 'base-64';
import queryString from 'query-string';
import { ICostsConfigSaveDataPayload } from '@/api-types';

/* configuration params for staging DB */
const { STAGING_DB_USERNAME, STAGING_DB_PASSWORD, STAGING_DB_REST_API_URL } =
  process.env;

async function getCostsConfig(
  query: Partial<{
    [key: string]: string | string[];
  }>
) {
  /* setup GET request options with basic auth */
  const requestOptions = {
    method: 'GET',
    headers: new Headers({
      'Content-Type': 'application/json',
      Authorization: `Basic ${base64.encode(
        `${STAGING_DB_USERNAME}:${STAGING_DB_PASSWORD}`
      )}`,
    }),
  };

  /* construct query param string from passed query object */
  const queryParamString = queryString.stringify(query);

  /* fetch the costs config */
  return fetch(
    `${STAGING_DB_REST_API_URL}/api/costs_config?${queryParamString}`,
    requestOptions
  );
}

async function setCostsConfig(body: ICostsConfigSaveDataPayload) {
  /* setup POST request options with basic auth */
  const requestOptions = {
    method: 'POST',
    headers: new Headers({
      'Content-Type': 'application/json',
      Authorization: `Basic ${base64.encode(
        `${STAGING_DB_USERNAME}:${STAGING_DB_PASSWORD}`
      )}`,
    }),
    body: JSON.stringify(body),
  };

  /* send the costs config */
  return fetch(`${STAGING_DB_REST_API_URL}/api/costs_config`, requestOptions);
}

const api = {
  getCostsConfig,
  setCostsConfig,
};

export default api;
