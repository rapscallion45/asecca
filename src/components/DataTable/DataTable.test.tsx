import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { ICostsConfigData } from '@/lib/api/api-types';
import DataTable from './DataTable';
import dataTableColumns from '../CostsConfigTable/costsConfigTableColumns';
import costsConfigDataMock from '../../../__mocks__/costsConfigDataMock';
import { IDataTableColumn } from './types';

/* mock callback for get cell value */
const mockGetCellValueCallback = jest.fn(
  (rowIdx: number, column: IDataTableColumn) =>
    costsConfigDataMock.costs[rowIdx][column.key as keyof ICostsConfigData]
);

/* Data Table Unit Tests */
/* ===================== */
describe('Data Table', () => {
  it('Renders correctly', async () => {
    /* perform snapshot test */
    const tree = renderer
      .create(
        <DataTable
          name="testtable"
          columns={dataTableColumns}
          editableColLabels={['Collection']}
          rows={costsConfigDataMock.costs.map((cost: ICostsConfigData) => ({
            label: cost.name,
          }))}
          getCellValueCallback={mockGetCellValueCallback}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Values', () => {
    /* ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render expected values for passed data structure', async () => {
      /* Arrange */
      /* Act */
      render(
        <DataTable
          name="testtable"
          columns={dataTableColumns}
          editableColLabels={['Collection']}
          rows={costsConfigDataMock.costs.map((cost: ICostsConfigData) => ({
            label: cost.name,
          }))}
          getCellValueCallback={mockGetCellValueCallback}
        />
      );

      /* Assert - correct value should be displayed from test data */
      costsConfigDataMock.costs.forEach((cost) => {
        expect(screen.getByText(cost.application)).toBeInTheDocument();
        expect(screen.getByText(cost.name)).toBeInTheDocument();
      });
    });
  });

  describe('Error and Loading', () => {
    /* ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render loading spinner when loading prop set', async () => {
      /* Arrange */
      /* Act */
      render(
        <DataTable
          name="testtable"
          columns={dataTableColumns}
          editableColLabels={['Collection']}
          rows={costsConfigDataMock.costs.map((cost: ICostsConfigData) => ({
            label: cost.name,
          }))}
          getCellValueCallback={mockGetCellValueCallback}
          isLoading
        />
      );

      /* Assert - loading spinner and message should be rendered */
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('Should render error message when loading prop set', async () => {
      /* Arrange */
      const errorMsg: string = 'Failed to laod data.';

      /* Act */
      render(
        <DataTable
          name="testtable"
          columns={dataTableColumns}
          editableColLabels={['Collection']}
          rows={costsConfigDataMock.costs.map((cost: ICostsConfigData) => ({
            label: cost.name,
          }))}
          getCellValueCallback={mockGetCellValueCallback}
          error={errorMsg}
        />
      );

      /* Assert - correct value should be displayed from test data */
      expect(screen.getByTestId('ErrorIcon')).toBeInTheDocument();
      expect(screen.getByText(errorMsg)).toBeInTheDocument();
    });
  });
});
