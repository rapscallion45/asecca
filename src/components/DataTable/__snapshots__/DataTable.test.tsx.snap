// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Table Renders correctly 1`] = `null`;
